text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,288 @@
+- Algolia:
+	- Algolia seems to use default word segmentation use user configurable separators. The use can also set the `queryLanguage` or provide a list of `stopWords`, but it is unclear how this work under the hood for languages such as Mandarin. They also word decompounging, meaning that they can split words that are mode of multiple words (mostly germanic/nordic languages apparently).
+
+### Explanation
+
+The tokenizer serves as a proxy for other tokenizers, specialized in the language detected by the Tokenizer. It is instantiated with a String, and is then polled for tokens until it's delepted:
+
+```rust
+let tokenizer = Tokenizer::new("The quick brown fox jumps over the lazy dog");
+let token_group = tokenizer.next();
+assert_eq!(token_group.tokens().map(|t| t.text()).collect(), ["The"]);
+assert_eq!(token_group.normalized().map(|t| t.text()).collect(), ["the"]);
+let token_group = tokenizer.next();
+assert_eq!(token_group.tokens().map(|t| t.text()).collect(), [" "])
+let token_group = tokenizer.next();
+assert_eq!(token_group.tokens().map(|t| t.text()).collect(), ["quick"])
+```
+
+#### highlight
+
+```rust
+fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
+  for (_key, value) in record.iter_mut() {
+    let old_value = mem::take(value);
+    let tokenizer = Tokenizer::new(&the_string);
+    for token_group in tokenizer {
+      // get the longest normalized token
+      let token = token_goup.normalized().max_by(|a, b| a.token_len().cmp(b.token_len()));
+      if token.kind() == TokenKind::Word {
+        let normalized_token = token.text();
+        let to_highlight = words.contains(&normalized_token);
+        if to_highlight { value.push_str("<mark>") }
+        value.push_str(token.original.text());
+        if to_highlight { value.push_str("</mark>") }
+      } else {
+        value.push_str(token);
+      }
+    }
+  }
+}
+```
+
+```rust
+// original
+fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
+    for (_key, value) in record.iter_mut() {
+        let old_value = mem::take(value);
+        for (token_type, token) in simple_tokenizer(&old_value) {
+            if token_type == TokenType::Word {
+                let lowercase_token = token.to_lowercase();
+                let to_highlight = words.contains(&lowercase_token);
+                if to_highlight { value.push_str("<mark>") }
+                value.push_str(token);
+                if to_highlight { value.push_str("</mark>") }
+            } else {
+                value.push_str(token);
+            }
+        }
+    }
+}
+```
```

Comment:
Could you explain a little bit more about the difference? This seems to not be possible on the current MeiliSearch. Because at the moment, the engine cannot give us all words (variations) that matched on documents. Is this a change to make? If yes, could you write it?